clash-cores/src/Clash/Cores/Xilinx/Ethernet/Gmii/Types.hs

```diff
@@ -0,0 +1,161 @@
+    bv2 = replaceBit (10 :: Int) linkSpeed0 bv1
+    bv3 = replaceBit (11 :: Int) linkSpeed1 bv2
+
+    -- Set link status
+    bv4 = setBit bv3 (15:: Int)
+
+
+-- | 10 Mbit and half duplex are the default values since they are the most likely
+-- to be supported by the link partner.
+instance Default AutoNegConfig where
+  def = AutoNegConfig {..}
+   where
+    cLinkSpeed = Speed10
+    cDuplexMode = HalfDuplex
+
+type ConfigVector = BitVector 5
+-- | A record representation of the configuration_vector.
+data Config = Config
+  { cUniDirectional :: Bool
+  , cLoopback :: Bool
+  , cPowerDown :: Bool
+  , cIsolateGmii :: Bool
+  , cAutoNegEnable :: Bool
+  } deriving (Generic, NFDataX, BitPack)
+
+-- | Transform the Config record to a configuration_vector.
+toConfigVector :: Config -> ConfigVector
+toConfigVector Config{..} = pack
+  ( cAutoNegEnable
+  , cIsolateGmii
+  , cLoopback
+  , cPowerDown
+  , cUniDirectional
+  )
+
+instance Default Config where
+  def = Config {..}
+   where
+    cUniDirectional = False
+    cLoopback = False
+    cPowerDown = False
+    cIsolateGmii = False
+    cAutoNegEnable = True
+
+type StatusVector = BitVector 16
+-- | A record representation of the status_vector.
+data Status = Status
+  { sLinkStatus :: Bool
+  , sLinkSynchronized :: Bool
+  , sRudiC :: Bool
+  , sRudiI :: Bool
+  , sRudiInvalid :: Bool
+  , sReceivedDisparityError :: Bool
+  , sReceivedInvalidCode :: Bool
+  , sPhyLinkStatus :: Bool
+  , sRemoteFault :: Maybe (BitVector 2)
+  , sLinkSpeed :: LinkSpeed
+  , sDuplexMode :: DuplexMode
+  , sPause :: Pause
+  } deriving (Generic, NFDataX, BitPack)
+
+-- | Transform the status_vector to a Status record.
+fromStatusVector :: StatusVector -> Status
+fromStatusVector v = Status{..}
+ where
+  -- There is no BitPack instance for 13-tuples without -flarge-tuples, so we break it up.
+  ( ( sLinkStatus
+    , sLinkSynchronized
+    , sRudiC
+    , sRudiI
+    , sRudiInvalid
+    , sReceivedDisparityError
+    )
+    , sReceivedInvalidCode
+    , sPhyLinkStatus
+    , remoteFaultCode
+    , linkSpeedBools
+    , duplexModeBool
+    , remoteFaultValid
+    , pauseBools) = unpack v
+  sRemoteFault = if remoteFaultValid then Just remoteFaultCode else Nothing
+  sLinkSpeed = case linkSpeedBools of
+    (True, False) -> Speed1000
+    (False, True) -> Speed100
+    (False, False) -> Speed10
+    _ -> errorX "fromStatusVector: invalid link speed"
+  sDuplexMode = if duplexModeBool then FullDuplex else HalfDuplex
+  sPause = case pauseBools of
+    (False, False) -> NoPause
+    (False, True) -> SymmetricPause
+    (True, False) -> AsymmetricPause
+    (True, True) -> SymmetricAsymmetricPause
+
+-- | A subset of the relevant output signals of the GMII to SGMII bridge.
+data BridgeOutput gmiiRx125 sgmii625 = BridgeOutput
+  { bridgeLvdsOut :: Signal sgmii625 Lvds
+  , bridgeGmiiRx :: Signal gmiiRx125 Gmii
+  , bridgeClk125 :: Clock gmiiRx125
+  , bridgeRst125 :: Reset gmiiRx125
```

Comment:
Is this reset synchronous or asynchronous? We probably need a constraint somewhere so it can only be used properly.

Reply:
The constraints have been added to the functions. Would you like to also put them on the record?

Reply:
No, I think it's good enough to have it just on the functions.